infra/build/functions/update_build_status.py

```diff
@@ -24,71 +24,128 @@
   credentials, image_project = google.auth.default()
   cloudbuild = build('cloudbuild',
                      'v1',
                      credentials=credentials,
                      cache_discovery=False)
 
+  history = []
+  last_build_status = None
+
   for build_id in reversed(build_ids):
     project_build = cloudbuild.projects().builds().get(projectId=image_project,
                                                        id=build_id).execute()
     if project_build['status'] == 'WORKING':
       continue
 
+    if not last_build_status:
+      last_build_status = project_build['status']
+
     if not builds_status.upload_log(build_id):
       log_name = 'log-{0}'.format(build_id)
       raise MissingBuildLogError('Missing build log file {0}'.format(log_name))
 
-    return project_build
+    success = builds_status.is_build_successful(project_build)
+
+    history.append({
+        'build_id': build_id,
+        'finish_time': project_build['finishTime'],
+        'success': success
+    })
 
-  return None
+    if len(history) == MAX_BUILD_LOGS:
+      break
 
+  return history, last_build_status
 
+
+# pylint: disable=too-many-locals
 def update_build_status(build_tag, status_filename):
   """Update build statuses."""
+  build_history = []
+  last_successful_build = []
   statuses = {}
   successes = []
   failures = []
+  not_yet_built = []
   for project_build in BuildsHistory.query(
-      BuildsHistory.build_tag == build_tag):
-    last_build = get_last_build(project_build.build_ids)
-    if not last_build:
-      logging.error('Failed to get last build for project %s',
-                    project_build.project)
+      BuildsHistory.build_tag == build_tag).order('project'):
+
+    history, last_build_status = get_build_history(project_build.build_ids)
+
+    if not last_build_status:
+      not_yet_built.append({'name': project_build.project, 'history': history})
       continue
 
-    if last_build['status'] == 'SUCCESS':
+    if last_build_status == 'SUCCESS':
       statuses[project_build.project] = True
-      successes.append({
-          'name': project_build.project,
-          'build_id': last_build['id'],
-          'finish_time': last_build['finishTime'],
-          'success': True,
-      })
+      successes.append({'name': project_build.project, 'history': history})
     else:
       statuses[project_build.project] = False
-      failures.append({
-          'name': project_build.project,
-          'build_id': last_build['id'],
-          'finish_time': last_build['finishTime'],
-          'success': False,
-      })
-
-  builds_status.upload_status(successes, failures, status_filename)
+      failures.append({'name': project_build.project, 'history': history})
+
+  for project in failures:
+    build_history.append(project)
+    last_success = ndb.Key(LastSuccessfulBuild,
+                           project.name + '-' + build_tag).get()
+    if not last_success:
+      continue
+    last_successful_build.append({
```

Comment:
this doesn't catch cases where we get a history like "S S F" (more recent on the right) where both the successes are newer than the last recorded successful build right? This should be handled, and it's possible some bug/problem may prevent this function from being run for some time.

I would suggest simplifying these 3 loops by folding everything into a single one

(pseudocode)
```python
  for project_build in BuildsHistory.query(
      BuildsHistory.build_tag == build_tag).order('project'):

   history = get_build_history(...)
   last_successful_build = update_last_success(project_build.project, history). # find latest success and update entity accordingly. 

   results['projects'].append({
      'name': project_name,
      'history': history,
      'last_successful_build': last_successful_build,
   }
```